### PR TITLE
472963  philips hue binding unsupported light

### PR DESCRIPTION
**Pull request by markusmazurczak**
Changed Thing usage in that way that it is not anymore necessary to
include specific bulb types as single Thing-Types. From now on there are
4 generic thing types that are supported (Dimmable light, Color light,
Extended Color light, Color temperature light) with there respective
channels. The Thing type to use for a discovered bulb is now determined
by it's type.

I normalize the light type before the supportedThingTypes-check is done

### CODE REVIEW

**Comment by kaikreuzer** [comment]
There is something mixed up with your commits - you have 5, including the one for the Osram bug.
But never mind, you can cleanup once we are ready to merge. Will try to provide you feedback tomorrow!

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java [hunk 1]
```diff
@@ -87,8 +86,8 @@ public void initialize() {
                 if (fullLight != null) {
                     updateProperty(Thing.PROPERTY_FIRMWARE_VERSION, fullLight.getSoftwareVersion());
                     // Check for Osram PAR16 50 TW bulb
-                    String thingTypeId = HueBindingConstants.THING_TYPE_PAR16_50_TW.getId();
-                    if (thingTypeId.equals("PAR16_50_TW")) {
+                    String thingUID = fullLight.getModelID();
+                    if (thingUID.equals("PAR16 50 TW")) {
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java]: Shouldn't that string moved to HueBindingConstants?

File: extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java [hunk 2]
```diff
@@ -47,15 +46,15 @@
  * @author Thomas Höfer - added thing properties
  * @author Jochen Hiller - fixed status updates for reachable=true/false
  * @author Markus Mazurczak - added code for command handling of OSRAM PAR16 50 bulbs
+ *         <<<<<<< HEAD
+ * @author Markus Mazurczak - Changed to generic thing types
+ *         =======
+ *         >>>>>>> refs/remotes/upstream/master
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueLightHandler.java]: You checked in a merge conflict...

**Comment by markusmazurczak** [comment]
Something is really messed up with that pull request :-)
Is it possible that you reject my request and I will create a new branch, cherry pick the actual changes and create a new pull request.

I am not able squash my commits :-(

**Comment by maggu2810** [comment]
No need to create a new request...
If you get a clean new branch, just force push that branch to this one and this PR is changed magically.

**Comment by markusmazurczak** [comment]
Alright, thx. I will create a clean branch and put that PAR16 type into BindingConstants

**Comment by markusmazurczak** [comment]
I think the commit is clean now :-)
Testsuite runs successfully and as you mentioned, maggu2810, I put the PAR16-Type ID as a constant into HueBindingConstants

**Comment by kaikreuzer** [comment]
> I think the commit is clean now :-)

Yes, but the PR still contains the old 3 commits as well...

One question regarding the implementation: Do you see any simple way to also allow backward compatibility? Some existing system can have persisted things like `hue:LCT001:123` and it would be nice if it continues to work without a manual migration. Furthermore, with the new implementation, we are effectively losing information about the exact device (all the specific information like labels & descriptions that we have in the XMLs).

So imho the ideal implementation would be the best of both worlds: Use the specific thing type, if there is one and otherwise fall back to the generic thing type. This actually just reminds me of the discussion about decoupling the Thing UID from its type (so that this can potentially be switched during the lifetime of the thing), see https://www.eclipse.org/forums/index.php?t=msg&th=1072347&goto=1716311&#msg_1716311.
The concept of having a generic handler as a fallback if no specific handler is available is something we will need at many places (I see it e.g. for Z-Wave and Zigbee as well), so the hue binding could actually be a good reference implementation for such a concept.
Sorry, if this means more work for this PR (or rather no direct merge), but I have the feeling that it would be worth to follow-up on this idea. What do others like @maggu2810, @cdjackson and @sjka think?

**Comment by markusmazurczak** [comment]
> Do you see any simple way to also allow backward compatibility?

I think that especially for the Hue binding the implementation of a fallback would be a pretty easy task. I  would implement a 'fallback' check in that private getThingUID method. If no Thing is found by it's model ID I can check for the bulb type.

This will just be an pretty Hue specific implementation. If some more general/generic Thing-Fallback mechanism is needed I need a bit more time of thinking about this ;).

One last question regarding
>but the PR still contains the old 3 commits as well...

Any hints for me how I can get rid of this 3 commits. I cannot squash the commits and the last thing I've done was creating a fresh new branch and cherry picked my changes. After that I did a force push and expected that everything will be as needed.

**Comment by sjka** [comment]
@kaikreuzer I'm not really sure if our idea with the generic ThingHandler really applies here, as the discovery service already can decide upon the final ThingType. It uses the more generic ones that @markusmazurczak introduces here and I don't see a scenario where the TingType would change later on. 
However, for the migration of existing things, maybe that could help us here. If we would leave the old ThingHandler as it is, and introduce a new one next to it which would handle the new types, then we could have both, the old specific ones for existing things and the new generic ones for newly added things in parallel. As this might be not ideal, we could actually teach the old ThingHandler how to migrate things to the new types, and by converting them literally handing them over to the new thing handler. They would keep the reference to their old ThingType in their UID though, which I'm not yet really sure if that's actually a good idea...

**Comment by maggu2810** [comment]
Do we need such a mechanism (update existing things) if we add a channel to a thing type (perhaps caused by a new firmware version, a method of an old API that has been made public ...), too?
So we should thing about.

I have read the forum post (Z-Wave, thing UID, ...) as the posts has been done. I do not see any limitation about using more generic IDs and set the model etc. in properties. Are there any ones?

File: extensions/binding/org.eclipse.smarthome.binding.hue.test/src/test/groovy/org/eclipse/smarthome/binding/hue/test/HueLightHandlerOSGiTest.groovy [hunk 3]
```diff
@@ -428,7 +426,7 @@ class HueLightHandlerOSGiTest extends OSGiTest {
             assertBridgeOnline(hueLightHandler.getBridge())
             enableHueLightChannels(hueLight)
             //Run the initialize method in case of PAR16 50 TW bulb testcase
-            if(currentState.modelId == 'PAR16 50 TW') {
+            if(currentState.modelId.equals(OSRAM_PAR16_MODELID)) {
```
Comment (dominicdesu) [inline extensions/binding/org.eclipse.smarthome.binding.hue.test/src/test/groovy/org/eclipse/smarthome/binding/hue/test/HueLightHandlerOSGiTest.groovy]: Can you remove this condition and always call the initialize() method? I don't think there is a reason to do it only for the OSRAM light.

Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.hue.test/src/test/groovy/org/eclipse/smarthome/binding/hue/test/HueLightHandlerOSGiTest.groovy]: FTR: This code has been meanwhile fixed by https://github.com/eclipse/smarthome/pull/687.

**Comment by markusmazurczak** [comment]
Let me try to give a little summary about what has to be done.

We want to support the "old" ThingTypes and also introduce the 4 generic thing types. This is to further support already discovered things so that users do not have to reconfigure there stuff.

If my understanding is correct what do you think about the following?
I implement the 4 generic types and also implement some type of migration for the "old" things in a way that the old UID is used furthermore.
As @maggu2810 already mentioned we use the setProperty Method of Thing to save/migrate the vendor and modelID.

**Comment by sjka** [comment]
Sounds like a good summary to me. 
I would be okay with the first part, i.e. that you implement the 4 generic ThingTypes. Regarding the migration, I'm currently working on implementing the change as suggested in 
https://www.eclipse.org/forums/index.php?t=msg&th=1072347&goto=1716311&#msg_1716311. That would allow us to optionally not include the ThingType in the ThingUID anymore, which would make migration between the different ThingTypes easier for you. I hope to have some results somewhen next week. If you have the patience, it would be great to test-drive this with the hue binding together then. Would that be okay for you?

**Comment by markusmazurczak** [comment]
> Would that be okay for you?
Sounds good to me

**Comment by dominicdesu** [comment]
Might be relevant to this PR: Seems that the latest firmware of the Hue Bridge does not support uncertified third party lights anymore: http://www.developers.meethue.com/content/friends-hue-program-update

**Comment by kaikreuzer** [comment]
Yes, I read about that yesterday. That's no good news, but let's hope that many devices will soon make it again as a "friend of hue".

**Comment by markusmazurczak** [comment]
Yeah, bad news. Hopefully they will rethink this a little with some type of information like "You are using an unsupported light at your own risk"

**Comment by markusmazurczak** [comment]
Good news from Philips :-)
http://www.developers.meethue.com/content/friends-hue-program-update?page=1#comment-1157

**Comment by markusmazurczak** [comment]
Just as a little marker for myself:

Will start with that after merge of #768 and #754

**Comment by sjka** [comment]
If you like, I'd be happy to get some feedback on #768 - especially about the [migration steps in the ThingManager](https://github.com/eclipse/smarthome/pull/768/files#diff-24d71ac793f6de6c8557ea06292497ddR678). Do you see any issues with moving things over from your newly introduced types to a more specific one like it's done there?

**Comment by kaikreuzer** [comment]
@sjka I have actually not yet fully understood the idea of the thing type migration.
If we have handlers for the specific and generic thing types: Why would there be a need for any migration?
It imho only makes sense if we declare the specific types deprecated and we ask people, not to use them anymore. In such a case they should also not be listed in the Paper UI anymore when manually creating a new thing.
What will happen if the handler tries a migration of a thing defined in a things-file? This is obviously not possible and the code should take care of this.

**Comment by markusmazurczak** [comment]
@kaikreuzer I dont understand why it is not possible to migrate things which are defined in an things-file. In my understanding, by now, at every start the thing-file defined specific things will be migrated everytime because the thing files are still read at startup.

**Comment by kaikreuzer** [comment]
FYI: https://github.com/eclipse/smarthome/pull/768 has finally been merged!
@sjka & @markusmazurczak How do you want to proceed with this PR now?

**Comment by sjka** [comment]
As we can now change thing types, I would propose the following:
* add the new "generic" thing types (as it is currently the case in this PR)
* keep the old, specific thing types including their handler (i.e. revert the deletions). Deleting thing type is a breaking change for everybody who currently has some hue devices paired
 * we can mark the specific thing types as ```listed="false"``` as soon as #1101 is merged
* let the new, generic thing handler use the changeThingType feature to switch over to the special thing types either (I'm undecided here)
 * always whenever a special thing type exists
 * only if makes sense because the specialized one has some additional capabilities

What do you think?

**Comment by markusmazurczak** [comment]
@sjka So you mean that we just use the new generic thing handler just in case someone binds a new thing and switch over to the old one as long as there is an existing binding?

**Comment by sjka** [comment]
Yes, that would be one option. The other one would be to not switch, unless there is a good reason to do so (currently I'm not aware of any such thing type where this would make sense, but who knows what Philips comes up with in future...). As I said, I don't have a strong opinion here.

The only hard requirement from my side is that we stay compatible and don't break existing installations.

**Comment by kaikreuzer** [comment]
I think whenever we have a specific thing type and we can identify the bulb as such, we should use the specific type during discovery - it gives the chance to provide more detailed information about it.
I also don't see any urgent need to dynamically switch from generic to specific or vice versa during runtime for now. So this can clearly be a phase 2.

**Comment by kaikreuzer** [comment]
@markusmazurczak Are you working on this or shall @sjka go for it instead?

**Comment by markusmazurczak** [comment]
@kaikreuzer Will look for it in about 1 or 2 weeks. I am pretty busy right know. If this is too long for you I will be fine if @sjka takes care about it

**Comment by kaikreuzer** [comment]
I guess @sjka did not find the time any earlier either. So if you, @markusmazurczak would go for it, it would be great!

**Comment by kaikreuzer** [comment]
@markusmazurczak Any news here?

**Comment by afuechsel** [comment]
Looks a little bit abandoned here. I would like to finalize this PR. To summarize: 

- specific thing-types remain as they are
- generic thing-types to be added
-- what about the fifth one: On/off light (ZigBee Device ID: 0x0000)? 
-- alert channel will be added to all generic thing-types
-- effect channel will be added to color, color_temperature and extendend-color-lights
- existing HueLightHandler can handle all thing types
- HueLightDiscoveryService to be changed? Should it discover generic thing-types as proposed in this PR or remain as it is and discover specific thing-types?

**Comment by kaikreuzer** [comment]
Superseded by https://github.com/eclipse/smarthome/pull/1652.

**Pull request closed without merging**